docs/design/configs.md

```diff
@@ -0,0 +1,97 @@
+# Generic Config Object
+
+## Abstract
+
+A generic ``Config`` object to store command-line flags and environment variables
+that is used for configuration of ``Kubernetes`` components living on top of ``etcd``
+is proposed in this proposal.
+
+## Problem
+
+Currently command-line flags and environment variables are used for component configurations in ``Kubernetes``.
+But, this approach is not suitable for a distributed computing environment. Specifically the 
+problems with the current approach are:
+
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+* Hard to change in different versions that may support different configuration options.
+
+## Solution
+
+A new ``Config`` resource is proposed to solve the aforementioned problems 
```

Comment:
As much as I dislike it, Kubernetes components use command-line flags for configuration. One thing this should do is make it possible for components not run as pods (which eventually there shouldn't be any) to retrieve configuration via the API.

Another key feature we'll need is an easy way to generate a configuration object, similar to the discussion around secrets in #4822.